### PR TITLE
Reduce size of `cc::Build`  and size of generated targets

### PR DESCRIPTION
**Pull request by NobodyXu**

### CODE REVIEW

File: src/target.rs [hunk 1]
```diff
@@ -2,129 +2,52 @@
...
+mod parser;
+
+pub(crate) use parser::TargetInfoParser;
 
 /// Information specific to a `rustc` target.
 ///
 /// See <https://doc.rust-lang.org/cargo/appendix/glossary.html#target>.
 #[derive(Debug, PartialEq, Clone)]
-pub(crate) struct TargetInfo {
+pub(crate) struct TargetInfo<'a> {
```
Comment (NobodyXu) [inline src/target.rs:20]: cc @madsmtm This is the optimization I was proposing, can you take a look and confirm this looks good please?

File: src/target/parser.rs [hunk 2]
```diff
@@ -0,0 +1,130 @@
+use std::{env, str::FromStr};
+
+use crate::{
+    target::{llvm, TargetInfo},
+    utilities::OnceLock,
+    Error, ErrorKind,
+};
+
+#[derive(Debug)]
+struct TargetInfoParserInner {
```
Comment (madsmtm) [inline src/target/parser.rs:10]: Nit:
```suggestion
struct ParsedTargetInfo {
```

(Or `OwnedTargetInfo`)

**Review by madsmtm — COMMENTED** [review]
Don't have a strong preference for either version (in favour is that matching on `&str` is nicer, against is that we have to maintain the list of struct fields).

File: src/target/parser.rs (hunk 2 above)
Comment (NobodyXu) [reply src/target/parser.rs:10]: Hmmmm I think `Inner` suffix is ok, if we want to rename it we probably want to rename the other one as well.

**Comment by NobodyXu** [comment]
> against is that we have to maintain the list of struct fields

True we need that extra field maintenance, but that would make the generated binary smaller.

AFAIK `std::borrow::Cow<'static, str>` is 4 pointers long, and `&str` is only 2 pointers long.

Given that we use `Cow` a lot it's almost 2x smaller constant data, so I think it should be worth the extra effort.

**Comment by NobodyXu** [comment]
P.S. I think that some crates use cc-rs at runtime as a normal dep, so this should reduce their binary size.

**Pull request merged**